File: tests/fixtures/pure_probe.mjs
#!/usr/bin/env node
/**
 * Candidate half of the pure-function corpus probe — see tests/test_pure_function_parity.py.
 *
 * Same protocol as its Python twin: a JSON job file in, one JSON document out. It calls the
 * exported functions and does nothing else — every branch it reaches is the port's own,
 * which is the only reason a probe is allowed to exist beside a byte matrix.
 */
import { readFileSync } from 'node:fs';

import path from 'node:path';

import {
  accentFor, defaultTheme, manifestIsClaude, statusLines, versionVerdict,
} from '../../js/inspect/status.mjs';
import { fenceFor, setupBuildArgs } from '../../js/build/generate.mjs';
import { cmpKey } from '../../js/inspect/diff.mjs';
import { descBlockProblem, isVendoredPath, validateIsVendored } from '../../js/hosts/native.mjs';
import { proseMirrorProblems, romanToInt } from '../../js/inspect/checks-authoring.mjs';
import { themesToCheck } from '../../js/inspect/doctor.mjs';
import { capitalize } from '../../js/hosts/installs.mjs';
import { writeText } from '../../js/lib/fs.mjs';
import { jsonDumpsCompact, parseJson } from '../../js/lib/json.mjs';
import {
  comparePaths, normcase, isAbsolutePath, toPlatformPath, which,
} from '../../js/lib/paths.mjs';
import {
  parseIntStrict, codePointLength, padEndToWidth, stripWhitespace, percentDecode,
} from '../../js/lib/text.mjs';
// `expanduser` and `resolvePath` are NOT in `fs.mjs` — they are host-config resolvers and
// live beside the config-dir lookups that need them. Verified rather than assumed; the
// private copies in `js/hooks.mjs` are a second pair and not these.
import { expanduser, resolvePath } from '../../js/hosts/hosts.mjs';
import { installedDefaults } from '../../js/hosts/installs.mjs';
import { installAgentEntryOf } from '../../js/maintain/uninstall.mjs';
import {
  ask, askChoice, collectSetupLines, confirm, javaMajorOk, modeOptions, postureOptions,
  promptLine, setupSummaryLines, themeOptions,
} from '../../js/maintain/setup.mjs';
import { unifiedDiff, splitLines } from '../../js/lib/udiff.mjs';
import { stampMinute } from '../../js/web/api.mjs';
import { buildPlan, daemonArgs, restartArgs } from '../../js/web/server.mjs';
import {
  fetchPhases, parseOrigin, countOccurrences, redactUrlCreds,
} from '../../js/maintain/update.mjs';
import {
  sliceSection, slugifyHeading, stripHarnessBlocks,
} from '../../js/web/docs.mjs';
import { webFirstOk } from '../../js/ui/menu.mjs';
import {
  clamp, detailLines, dwidth, fit, glyphs, icon, logoLines, mark, progressBar, spin,
  themeFlair, themePreview, truncd, tuiEntries,
} from '../../js/ui/tui.mjs';
import { animOk, artFor, height, place, playLine, tile } from '../../js/ui/anim.mjs';
import { makeSandbox } from '../helpers/sandbox.mjs';

/**
 * P7c. `anim_ok` and `play_line` need two seams the reference probe also takes: the
 * environment (read at CALL time by `_anim_ok`, so a per-case dict is the real read and not
 * a monkeypatch) and stdout.
 *
 * THE CAPTURE REPLACES `process.stdout.write`, NOT `process.stdout`. `printOut` closes over
 * `process.stdout` at call time, so patching the method is enough — and it has to be the
 * method, because the probe's own results document goes out on the SAME stream afterwards.
 * `isTTY` is assigned rather than faked with a device: `< /dev/null` reads as a TTY to
 * Python on Windows, which is how an earlier session ran a whole wizard against a live
 * install.
 */
function withEnvAndStdout(env, tty, fn) {
  const keys = ['GENESEED_NO_ANIM', 'GENESEED_TUI_PLAIN', 'TERM', 'COLUMNS'];
  const saved = Object.fromEntries(keys.map((k) => [k, process.env[k]]));
  const realWrite = process.stdout.write.bind(process.stdout);
  const realTty = process.stdout.isTTY;
  let buf = '';
  for (const k of keys) delete process.env[k];
  for (const [k, v] of Object.entries(env)) if (v !== null) process.env[k] = v;
  process.stdout.isTTY = tty;
  process.stdout.write = (chunk) => { buf += chunk; return true; };
  try {
    const value = fn();
    return value === undefined ? buf : value;
  } finally {
    process.stdout.write = realWrite;
    process.stdout.isTTY = realTty;
    for (const k of keys) {
      if (saved[k] === undefined) delete process.env[k];
      else process.env[k] = saved[k];
    }
  }
}

/** One scratch dir per process for `write_text_linesep`, made on first use. */
let SCRATCH = null;
const scratchFile = () => {
  if (SCRATCH === null) SCRATCH = makeSandbox('pure-probe-').path;
  return path.join(SCRATCH, 't.txt');
};

const FNS = {
  version_verdict: (a) => versionVerdict(a[0], a[1]),
  status_lines: (a) => statusLines(a[0], a[1]),
  manifest_is_claude: (a) => manifestIsClaude(a[0]),
  accent_for: (a) => accentFor(a[0]),
  default_theme: () => defaultTheme(),
  py_len: (a) => codePointLength(a[0]),
  py_ljust: (a) => padEndToWidth(a[0], a[1]),
  fence_for: (a) => fenceFor(a[0]),
  unified_diff: (a) => unifiedDiff(a[0], a[1], {
    fromfile: 'source/f', tofile: 'deployed/f', lineterm: '',
  }),
  py_split_lines: (a) => splitLines(a[0]),
  cmp_key: (a) => cmpKey(a[0], a[1]),
  py_capitalize: (a) => capitalize(a[0]),
  setup_build_args: (a) => setupBuildArgs(...a),
  themes_to_check: (a) => themesToCheck(...a),
  roman_to_int: (a) => romanToInt(a[0]),
  desc_block_problem: (a) => descBlockProblem(a[0]),
  // `skill_stems` is a `set` on the reference side and a `Set` here; JSON carries a list.
  prose_mirror_problems: (a) => proseMirrorProblems(a[0], a[1], a[2], new Set(a[3]), a[4]),
  is_vendored_path: (a) => isVendoredPath(a[0]),
  validate_is_vendored: (a) => validateIsVendored(a[0]),
  py_which: (a) => which(a[0], a[1]),
  py_is_absolute: (a) => isAbsolutePath(a[0]),

  // ---- the primitives whose answers ARE the bytes on a user's disk --------------------
  py_path_str: (a) => toPlatformPath(a[0]),
  py_resolve: (a) => resolvePath(a[0]),
  // `toPlatformPath` AFTER IT, because the reference's `expanduser` is a Path METHOD and this
  // one is a string function: `Path("")` is `.` before `expanduser` is even consulted, so a
  // raw comparison would score `py_path_str`'s job as an `expanduser` divergence. The
  // composite is what `js/hooks.mjs` actually writes, and `py_path_str` gates its own half
  // over its own corpus — so a real tilde divergence still shows through, and does.
  expanduser: (a) => toPlatformPath(expanduser(a[0])),
  normcase: (a) => normcase(a[0]),
  compare_paths: (a) => comparePaths(a[0], a[1]),
  json_dumps: (a) => jsonDumpsCompact(a[0]),
  parse_json_roundtrip: (a) => jsonDumpsCompact(parseJson(a[0])),
  py_strip_space: (a) => stripWhitespace(a[0]),
  // HEX, NOT TEXT — see the reference probe's comment. `readFileSync` with no encoding is
  // the raw buffer, which is the only reading that can see `writeText`'s translation.
  write_text_linesep: (a) => {
    const p = scratchFile();
    writeText(p, a[0]);
    return readFileSync(p).toString('hex');
  },
  install_agent_entry_of: (a) => installAgentEntryOf(a[0]),
  py_int: (a) => parseIntStrict(a[0]),
  java_major_ok: (a) => javaMajorOk(...a),
  theme_options: () => themeOptions(),
  posture_options: () => postureOptions(),
  mode_options: () => modeOptions(),
  installed_defaults: () => installedDefaults(),
  // Seconds in, as `st_mtime` is; `stampMinute` takes ms, as `mtimeMs` is.
  minute_stamp: (a) => stampMinute(a[0] * 1000),
  py_unquote: (a) => percentDecode(a[0]),
  build_plan: (a) => buildPlan(a[0], a[1], a[2], a[3]),
  daemon_args: (a) => daemonArgs(a[0], a[1]),
  restart_args: (a) => restartArgs(a[0]),
  slugify_heading: (a) => slugifyHeading(a[0]),
  // `tuple` on the reference side, a two-element list over JSON.
  slice_section: (a) => sliceSection(a[0], a[1]),
  strip_harness_blocks: (a) => stripHarnessBlocks(a[0], a[1]),
  setup_summary_lines: (a) => setupSummaryLines(...a),
  // The stdin readers. Their PROMPTS are stdout, which is why the wizard job is compared as
  // raw bytes rather than through a JSON parse — see the test's docstring.
  ask: (a) => ask(...a),
  prompt_line: (a) => promptLine(a[0]),
  confirm: (a) => confirm(...a),
  ask_choice: (a) => askChoice(a[0], a[1], a[2]),
  collect_setup_lines: () => collectSetupLines(),
  // `_update`'s pure four (P8a). `parseOrigin` returns a namedtuple on the reference side,
  // which JSON carries as a list — hence the two-element array here.
  parse_origin: (a) => { const o = parseOrigin(a[0]); return [o.url, o.githubSlug]; },
  redact_url_creds: (a) => redactUrlCreds(a[0]),
  count_or_zero: (a) => countOccurrences(a[0]),
  fetch_phases: (a) => fetchPhases(a[0], a[1]),
  // P7a. `args[0]` FAKES `isTTY`, because a probe's stdin is a seeded file and every branch
  // past the isatty test would otherwise be dead on both sides — the one input no cell and
  // no ordinary corpus entry can vary. `process.stdin.isTTY` is a plain property here, as
  // `sys.stdin` is a plain module attribute there; both are put back afterwards.
  web_first_ok: (a) => {
    const real = process.stdin.isTTY;
    process.stdin.isTTY = a[0];
    try { return webFirstOk(); } finally { process.stdin.isTTY = real; }
  },

  // ---- P7b: the TUI's pure half ------------------------------------------------------
  glyphs: (a) => glyphs(a[0]),
  dwidth: (a) => dwidth(a[0]),
  truncd: (a) => truncd(a[0], a[1]),
  fit: (a) => fit(a[0], a[1]),
  icon: (a) => icon(a[0]),
  mark: (a) => mark(a[0]),
  spin: (a) => spin(a[0]),
  logo_lines: () => logoLines(),
  clamp: (a) => clamp(a[0], a[1], a[2]),
  progress_bar: (a) => progressBar(...a),
  theme_preview: (a) => themePreview(a[0]),
  theme_flair: (a) => themeFlair(a[0]),
  tui_entries: (a) => tuiEntries(a[0]),
  detail_lines: (a) => detailLines(a[0], a[1], a[2]),
  // The exhaustive one — see the reference probe's comment. `String.fromCodePoint` is the
  // codepoint-wise constructor; `String.fromCharCode` would build a UTF-16 unit and answer
  // for the wrong character everywhere above U+FFFF, which is where the emoji rule lives.
  // ---- P7c: `theme_anim`, the line mode's install animation ---------------------------
  art_for: (a) => {
    const got = artFor(a[0]);
    return {
      title: got.title, sprite: got.sprite, ground: got.ground || '', card: got.card,
    };
  },
  anim_place: (a) => place(a[0], a[1], a[2]),
  anim_tile: (a) => tile(a[0], a[1], a[2]),
  anim_height: (a) => height(a[0]),
  anim_ok: (a) => withEnvAndStdout(a[1], a[0], () => animOk()),
  play_line: (a) => withEnvAndStdout(a[3], a[2], () => { playLine(a[0], a[1]); }),

  dwidth_rle: (a) => {
    const runs = [];
    let prev = null;
    for (let cp = a[0]; cp < a[1]; cp += 1) {
      const w = dwidth(String.fromCodePoint(cp));
      if (w !== prev) { runs.push([cp, w]); prev = w; }
    }
    return runs;
  },
};

const job = JSON.parse(readFileSync(process.argv[2], 'utf8'));
const results = job.cases.map((c) => {
  const fn = FNS[c.fn];
  if (!fn) throw new Error(`pure_probe: unknown fn '${c.fn}'`);
  return fn(c.args);
});
// Plain `JSON.stringify`, and the PYTHON side is the one that moved: P5i's wizard job
// compares the probe's whole stdout BYTE FOR BYTE, so the serialiser became part of what is
// compared, and `json.dumps`' defaults (`', '` / `': '`, `ensure_ascii=True`) are not
// `JSON.stringify`'s. `jsonDumpsCompact` is the helper that reproduces them and it REFUSES a
// bare number by design — the int-vs-float distinction it protects is real, and half these
// probes return numbers. So the probes meet at `separators=(',', ':')` + `ensure_ascii=False`
// instead, which both languages can spell natively.
process.stdout.write(JSON.stringify({ results }));

File: tests/fixtures/pure_probe.py
#!/usr/bin/env python3
"""Reference half of the pure-function corpus probe — see tests/test_pure_function_parity.py.

Reads a JSON job `{"cases": [{"fn": ..., "args": [...]}, ...]}` from argv[1] and writes
`{"results": [...]}` to stdout as UTF-8. One process per environment, because
`_harness_tui_draw._TUI_ASCII` is read at IMPORT time: switching it inside a process is
monkeypatching, and running a second process is the same thing the real CLI does.
"""
from __future__ import annotations

import datetime
import difflib
import json
import shutil
import sys
import urllib.parse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "rituals"))
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))
import harness  # noqa: E402


def run(fn: str, args: list):
    # `_update`'s pure four (P8a). Imported lazily so a probe that never asks for them does
    # not pay `import build` twice — `harness` above already did.
    # P7a. `args[0]` FAKES `sys.stdin.isatty()`, because a probe's stdin is a seeded file and
    # every branch past the isatty test would otherwise be dead on both sides — the one input
    # no cell and no ordinary corpus entry can vary. A stand-in object rather than a real
    # terminal: there is no pty on Windows, and `< /dev/null` is reported as a TTY by Python
    # here, which is how a previous session ran an entire wizard against a live install.
    if fn == "web_first_ok":
        class _Stdin:
            def __init__(self, tty): self._tty = tty
            def isatty(self): return self._tty
        real, sys.stdin = sys.stdin, _Stdin(args[0])
        try:
            return harness._web_first_ok()
        finally:
            sys.stdin = real
    if fn in ("parse_origin", "redact_url_creds", "count_or_zero", "fetch_phases"):
        sys.path.insert(0, str(Path(__file__).resolve().parents[2] / "rituals"))
        import _update
        if fn == "parse_origin":
            return list(_update._parse_origin(args[0]))
        if fn == "redact_url_creds":
            return _update._redact_url_creds(args[0])
        if fn == "count_or_zero":
            return _update._count(args[0])
        # `_fetch_streaming`'s reader loop, lifted out of its thread: the same lines in, the
        # ones it would LOG out, plus the phase it ends on. The loop is inline in the
        # reference (it closes over `p.stdout`, `lines` and `log`), so this reproduces it
        # rather than calling it — which is the one shape this corpus allows, and only
        # because the alternative is no gate at all on the branch that keeps a fetch's
        # thousand progress repaints out of the install log.
        logged, last = [], args[1]
        for raw in args[0]:
            line = _update._redact_url_creds(raw.strip())
            if not line:
                continue
            phase = line.split(":", 1)[0]
            if phase != last:
                logged.append(line)
                last = phase
        return [logged, last]
    if fn == "unified_diff":
        return list(difflib.unified_diff(args[0], args[1], fromfile="source/f",
                                         tofile="deployed/f", lineterm=""))
    if fn == "py_split_lines":
        return args[0].splitlines()
    if fn == "cmp_key":
        return harness._cmp_key(args[0], args[1])
    if fn == "py_capitalize":
        return args[0].capitalize()
    if fn == "setup_build_args":
        return harness._setup_build_args(*args)
    if fn == "version_verdict":
        return harness._version_verdict(*args)
    if fn == "status_lines":
        return harness._status_lines(args[0], args[1])
    if fn == "manifest_is_claude":
        return harness._manifest_is_claude(Path(args[0]))
    if fn == "accent_for":
        return harness._accent_for(args[0])
    if fn == "default_theme":
        return harness._default_theme()
    if fn == "py_len":
        return len(args[0])
    if fn == "py_ljust":
        return args[0].ljust(args[1])
    if fn == "fence_for":
        return harness._fence_for(args[0])
    if fn == "themes_to_check":
        return harness._themes_to_check(*args)
    if fn == "roman_to_int":
        return harness._roman_to_int(args[0])
    if fn == "desc_block_problem":
        return harness.build._desc_block_problem(args[0])
    if fn == "prose_mirror_problems":
        # `skill_stems` is a SET on this side and a Set on the other; JSON carries a list.
        return harness._prose_mirror_problems(args[0], args[1], args[2], set(args[3]), args[4])
    if fn == "is_vendored_path":
        return harness.build.is_vendored_path(Path(args[0]))
    if fn == "validate_is_vendored":
        return harness.build._validate_is_vendored(Path(args[0]))
    if fn == "py_which":
        return shutil.which(args[0], path=args[1])
    if fn == "py_is_absolute":
        return Path(args[0]).is_absolute()
    if fn == "py_int":
        # `int(s)`, or None where it raises. The corpus never pads: `int` strips whitespace
        # and `pyInt` deliberately does not, because both of its callers have already run
        # `str.strip()` — see its docblock.
        try:
            return int(args[0])
        except ValueError:
            return None
    if fn == "java_major_ok":
        return harness._java_major_ok(*args)
    if fn == "theme_options":
        return harness._theme_options()
    if fn == "posture_options":
        return harness._posture_options()
    if fn == "mode_options":
        return harness._mode_options()
    if fn == "installed_defaults":
        return harness._installed_defaults()
    if fn in ("slugify_heading", "slice_section", "strip_harness_blocks"):
        # `_web_docs`'s pure halves, reached through the facade so the cross-module
        # splice is the same one the server runs under. Imported lazily: `web` pulls
        # `build` in with it and every other case here is cheaper without that.
        import web  # noqa: PLC0415
        if fn == "slugify_heading":
            return web._slugify_heading(args[0])
        if fn == "slice_section":
            return list(web._slice_section(args[0], args[1]))
        return web._strip_harness_blocks(args[0], args[1])
    if fn in ("build_plan", "daemon_args", "restart_args"):
        # `_web_server`'s pure halves, reached through the facade for the reason the docs
        # ones below are: the cross-module splice is the same one the server runs under.
        import web  # noqa: PLC0415
        if fn == "build_plan":
            return web._build_plan(Path(args[0]), Path(args[1]), args[2], args[3])
        if fn == "daemon_args":
            return web._daemon_args(args[0], args[1])
        return web._restart_args(args[0])
    if fn == "py_unquote":
        return urllib.parse.unquote(args[0])
    if fn == "minute_stamp":
        # `_web_overview.api_overview`'s `build_time`, spelled as the reference spells it:
        # a NAIVE `fromtimestamp`, which is local time. `stamp_doctor`'s `checked_at` is
        # `time.strftime` of the same format on `localtime()`, so one corpus covers both.
        return datetime.datetime.fromtimestamp(args[0]).strftime("%Y-%m-%d %H:%M")
    if fn == "setup_summary_lines":
        return harness._setup_summary_lines(*args)
    # ---- the stdin readers. Their PROMPTS are stdout, which is why the wizard job is
    # compared as raw bytes rather than through `json.loads` — see the test's docstring.
    if fn == "prompt_line":
        # `input(prompt)`, with EOF told apart from an empty line — which is the whole
        # difference `_web_server.serve`'s npm prompt branches on. Spelled as two calls
        # because `input(prompt)` writes the prompt and then reads, which is what the twin
        # does explicitly; `input()` alone would leave the prompt out of the compared stdout.
        sys.stdout.write(args[0])
        try:
            return input()
        except EOFError:
            return None
    if fn == "ask":
        return harness._ask(*args)
    if fn == "confirm":
        return harness._confirm(*args)
    if fn == "ask_choice":
        return harness._ask_choice(args[0], [tuple(o) for o in args[1]], args[2])
    if fn == "collect_setup_lines":
        return harness._collect_setup_lines()
    if fn == "install_agent_entry_of":
        # `_install_agent_entry`'s project arm, minus the file read: the whole decision is
        # which `instructions` entry it picks, and the list is the input.
        for e in args[0]:
            if isinstance(e, str) and not Path(e).is_absolute() and Path(e).name == "AGENT.md":
                return e
        return "AGENT.md"

    # ---- P7b: the TUI's pure half ----------------------------------------------------
    #
    # Reached through `harness`, not through `_harness_tui_draw` directly, because the
    # display tier is a MODULE CONSTANT read at import time and `harness` is what imports
    # it — asking the submodule again would be a second import in some Python versions and
    # a different constant in none of them, but the caller's view is the one under test.
    if fn == "glyphs":
        return harness._glyphs(args[0])
    if fn == "dwidth":
        return harness._dwidth(args[0])
    if fn == "truncd":
        return harness._truncd(args[0], args[1])
    if fn == "fit":
        return harness._fit(args[0], args[1])
    if fn == "icon":
        return harness._icon(args[0])
    if fn == "mark":
        return harness._mark(args[0])
    if fn == "spin":
        return harness._spin(args[0])
    if fn == "logo_lines":
        return harness._logo_lines()
    if fn == "clamp":
        return harness._clamp(args[0], args[1], args[2])
    if fn == "progress_bar":
        return harness._progress_bar(*args)
    if fn == "theme_preview":
        return [list(r) for r in harness._theme_preview(args[0])]
    if fn == "theme_flair":
        return harness._theme_flair(args[0])
    if fn == "tui_entries":
        return [list(r) for r in harness._tui_entries(args[0])]
    if fn == "detail_lines":
        return harness._detail_lines(args[0], args[1], args[2])
    if fn == "dwidth_rle":
        # THE EXHAUSTIVE ONE. `_dwidth` is the only function in this phase with no Node
        # equivalent at any rung — `unicodedata.east_asian_width` and `unicodedata.
        # combining` have no counterpart in JavaScript, whose `\p{...}` escapes carry
        # neither property — so the port carries hand-written range tables, and a
        # hand-written Unicode table is right until it is not.
        #
        # So this case is not a sample of the input space, it IS the input space: every
        # codepoint in [lo, hi), run-length encoded. A run boundary appears wherever the
        # width changes, so two implementations agree if and only if these lists are equal,
        # and a range the port got wrong has no input left to hide behind.
        #
        # Measured against the LIVE `unicodedata` on this side rather than a table copied
        # from it — the day the reference's Unicode version moves under the port's tables,
        # this goes red and names the codepoint instead of agreeing with itself.
        lo, hi = args
        runs, prev = [], None
        for cp in range(lo, hi):
            w = harness._dwidth(chr(cp))
            if w != prev:
                runs.append([cp, w])
                prev = w
        return runs
    raise SystemExit(f"pure_probe: unknown fn {fn!r}")


def main() -> int:
    job = json.loads(Path(sys.argv[1]).read_text(encoding="utf-8"))
    out = [run(c["fn"], c["args"]) for c in job["cases"]]
    # FLUSH BEFORE DROPPING TO `.buffer`. The wizard cases print through `sys.stdout`, whose
    # TextIOWrapper buffers when stdout is a pipe; writing to the underlying binary buffer
    # without flushing first puts the results document AHEAD of the prompts that produced it.
    sys.stdout.flush()
    # `separators` + `ensure_ascii=False` so this document is BYTE-identical to the Node
    # probe's `JSON.stringify` — P5i's wizard job compares the probes' whole stdout rather
    # than parsing it, which put the serialiser itself under the gate. `json.loads` on the
    # reading side is indifferent to both.
    sys.stdout.buffer.write(json.dumps({"results": out}, separators=(",", ":"),
                                       ensure_ascii=False).encode("utf-8"))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
